### DISCUSSION TITLE
Update entity material color

### DISCUSSION THREAD

**Discussion by AlexandreFleutelot** [bevyengine/bevy#8487] [Q&A] [answered]
Hello,

I try to replicate the waving cube exemple from raylib with bevy (https://www.raylib.com/examples/models/loader.html?name=models_waving_cubes)

Here is the definition of my bevy project :

I have two component on each spawned cube :
- Cube //entity
- CubeGrid(f32,f32,f32) //cube position in the grid

Startup systems :
- spawn_cubes
- spawn_light_and_camera

Systems:
- update_camera_pos
- update_cubes_pos
- update_cubes_size
- update_cubes_color (?????)

Everything works fine apart from the update_cubes_color system.

My cube are Spawn with material defined like this:
material: materials.add(Color::hsl(0.8, 0.7, 0.6).into()),

How Can i update the color of each cube based on its CubeGrid coordonate in the update_cubes_color system?
(I already understood that I have two work with handles but du not success to implement it)

I know there are an other error (HSV ->HSL) but i will adress this one later as it seems much easier ;)

Thanks a lot,
Alexandre

**Comment by nicopap** [comment, marked as answer]
You need at least two system parameters to your `update_cube_color` system:

1. `mut materials: ResMut<Assets<StandardMaterial>>`
2. `handles: Query<(&CubeGrid, &Handle<StandardMaterial>)>`

Now, in `update_cube_color`, you need to access the `&mut StandardMaterial` for your cube's `Handle`:

```rust
let material = materials.get_mut(cube_handle).unwrap();
```
With that, you can update the materials' color with
```rust
material.base_color = color_from_grid(cube_grid);
```

You should be done.

### Notes

All your cubes need to have a distinct material for this approach to work, otherwise you would just be updating the same material multiple times. As you described your current program, it seems that each cube does indeed have a distinct material.

**Comment by AlexandreFleutelot** [comment]
Thanks, it solves my issue :)

Surprisingly the FPS is not very good (about 60 fps where raylib runs at about 700 fps). 
I was thinking this kind of scene would be a good exemple for an ECS but maybe something is wrong in my implementation.

If someone is interested in the complete project, you can find it in : 
https://github.com/AlexandreFleutelot/waving-cubes.git

**Reply by NiklasEi** [reply]
About 60 FPS sounds suspicious. Don't you have to set the `PresentMode` to `AutoNoVsync` to get accurate FPS?
